util: Add getinfo, parsing and naming functions for layered providers

a-ilango opened the pull request:
Also add some logging to fi_check_info

a-ilango commented:
@shefty @jithinjosepkl

shefty commented:
Please defer this until after 1.3

shefty commented on prov/util/src/util_attr.c:
```diff
@@ -321,11 +500,15 @@ int fi_check_info(const struct fi_provider *prov,
 	if ((user_info->mode & prov_info->mode) != prov_info->mode) {
 		FI_INFO(prov, FI_LOG_CORE, "needed mode not set\n");
+		FI_INFO(prov, FI_LOG_CORE, "Expected: %s\n", fi_tostr(&prov_info->mode, FI_TYPE_MODE));
+		FI_INFO(prov, FI_LOG_CORE, "Given: %s\n", fi_tostr(&user_info->mode, FI_TYPE_MODE));
```
There are several places where the line lengths look a little too long.  Please keep closer to 80-90 characters.

shefty commented on prov/util/src/util_attr.c:
```diff
@@ -56,11 +57,167 @@ static int fi_valid_addr_format(uint32_t prov_format, uint32_t user_format)
+	return ret;
+}
+
+int fi_check_name(char *user_name, char *prov_name, enum fi_check_type type)
```
From just the perspective of this call, the name check is whether the name is an exact match or a prefix match.  I would change the API to reflect this, rather than assume it based on the provider layering.

a-ilango replied:
To clarify, do you want this call to do an exact name match and if that fails, do a prefix match?

jithinjosepkl replied:
Can we just have these as two different APIs, instead of having the type as argument?

shefty replied:
I was thinking of something like fi_match_str(src, dst, match_type), where match_type = {exact, prefix, ...}.  However separate calls may be better, in which case you would just add fi_strcmp_prefix(), or such.

a-ilango replied:
Even if we had separate functions, the match_type argument would still have to waterfall from the calling function (i.e. fi_check_info -> fi_check_fabric_attr). So I think we could go with having match_type as an argument itself.

shefty commented on prov/util/src/util_attr.c:
```diff
@@ -56,11 +57,167 @@ static int fi_valid_addr_format(uint32_t prov_format, uint32_t user_format)
+
+		ret = strcasecmp(prov_name, tok[0]);
+		free(tok[0]);
+		return ret;
```
I didn't fully go through the parse_name call, but couldn't this be simplified to strncasecmp(user_name, prov_name, strlen(prov_name))?

a-ilango replied:
Yes. I'll update this to use strncasecmp.

shefty commented on prov/util/src/util_attr.c:
```diff
@@ -56,11 +57,167 @@ static int fi_valid_addr_format(uint32_t prov_format, uint32_t user_format)
+	if (ret)
+		return ret;
+	if (!(base_attr->name = strdup(tok[1])))
+		ret = -FI_ENOMEM;
```
Is this essentially doing a strndup?

a-ilango replied:
We do know domain_attr->name prefix for the layered provider. I think we could simplify this to use strdup where the string starting position is offset by strlen(prefix) + 1.

hppritcha commented:
This PR breaks gni provider build.
@sungeunchoi

hppritcha commented:
I opened a PR against this PR. @a-ilango please merge my PR in to your's before merging this.

a-ilango commented:
@hppritcha Thanks, I'll merge it with my PR. So travis-ci doesn't build gni provider?

hppritcha commented:
Unfortunately not.  I'll try again with Cray to allow for rpms or tarball to be installed for travis build that would allow for builds.  I could hook up jenkins if all else fails.

lanl-ompi commented:
bot:retest

lanl-ompi commented:
Test failed
 
Test FAILed.

lanl-ompi commented:
Test FAILed.

lanl-ompi commented:
Test failed
 
Test FAILed.

lanl-ompi commented:
Test FAILed.

a-ilango commented:
Updated the PR w.r.t the comments. @hppritcha I have updated the gni provider in this PR itself since some modifications were required.

shefty commented on include/fi_util.h:
```diff
@@ -378,5 +386,27 @@ void fi_fabric_insert(struct util_fabric *fabric);
+int utilx_tr_base_domain_attr(struct fi_domain_attr *base_attr, char *prefix,
+		struct fi_domain_attr *layer_attr);
+int utilx_tr_base_fabric_attr(struct fi_fabric_attr *base_attr, char *prefix,
+		struct fi_fabric_attr *layer_attr);
```
I would prefer a different prefix for these.  'utilx_' is a little long, but 'utilx_tr_base_' or 'util_tr_layer_' it's too much.  (I don't even know what the 'tr' stands for.)  And the two sets of calls take the same arguments, so it's unclear which call to make.

My vote for internal APIs that do not belong to any specific provider is 'ofi'.  Please look at the naming of the existing attr related calls and use either 'check' or 'alter' in the name, based on the behavior.  It may also be possible to modify the existing calls, rather than introducing new internal APIs.

shefty commented on prov/util/src/util_attr.c:
```diff
@@ -56,11 +71,148 @@ static int fi_valid_addr_format(uint32_t prov_format, uint32_t user_format)
+err:
+	free(name_dup);
+	return -FI_EOTHER;
+}
```
There's something about adding a name parsing function that rubs me the wrong way.  It seems like overkill and assumes some sort of format to a fabric name that is not necessarily true.  Why isn't a prefix check on a name sufficient?

a-ilango replied:
We need the parsing function since we have the provider name embedded in the fabric name.

shefty commented on prov/util/src/util_attr.c:
```diff
@@ -56,11 +71,148 @@ static int fi_valid_addr_format(uint32_t prov_format, uint32_t user_format)
+
+/* Construct fabric name for the layered provider in the following format:
+ * <layered_prov>_<base_prov_name>_<base_prov_fabric_name>
+ * e.g. rxm_verbs_IB-1234 */
```
Do we need to capture the base provider name?  It seems that the base fabric name should be sufficient.

a-ilango replied:
The base provider name needs to be captured in the fabric name (in fi_info returned to user) so that when the user passes the fi_info again (on an fi_domain call for example), we can correctly set the fabric_attr->prov_name when calling fi_getinfo for base provider.

shefty replied:
I don't think the fabric name should change at all.  We're still on the same fabric; it's just applying a different protocol.

shefty commented:
The code itself looks fine.  I just have some discussion related questions.

a-ilango commented:
Removed the parsing function (as we are not modifying the fabric name anymore) to replace it with simple functions for adding and removing string prefix for layered providers. Changed the function names. @shefty Let me know if this looks good.

hppritcha commented:
checking to see if the lanl-bot is doing better now.
bot:retest

shefty commented on prov/util/src/util_attr.c:
```diff
@@ -56,11 +71,79 @@ static int fi_valid_addr_format(uint32_t prov_format, uint32_t user_format)
+	return (type == FI_MATCH_PREFIX) ?
+		strncasecmp(user_name, prov_name, strlen(prov_name)) :
+		strcasecmp(prov_name, user_name);
+}
```
Nearly all of the fi_* functions return 0 on success.  I would negate the strcmp return values here to match that, rather than passing through the values.  Also declare as static.

a-ilango replied:
strcasecmp returns 0 on a match (success). So I think we are good here. I can update the function to static.

shefty replied:
> strcasecmp returns 0 on a match (success). So I think we are good here. I
> can update the function to static.

Oh yeah, duh...  :S

shefty commented:
One minor adjustment then I think we're good to merge.  Thanks

a-ilango commented:
Done!
